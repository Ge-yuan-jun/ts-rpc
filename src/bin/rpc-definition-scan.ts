import glob from 'glob'
import { Project, Node, ClassDeclaration, FunctionDeclaration, MethodDeclaration, InterfaceDeclaration, SyntaxKind, TypeReferenceNode, TypeAliasDeclaration, MethodSignature } from 'ts-morph'
import path from 'path'
import { existsSync } from 'fs'
import { IScanResult, TRPCMetaData } from '../interface'

export function scan (filePaths: string[]): IScanResult {
  const files = filePaths.map(f => glob.sync(f)).flat()
  const prj = new Project({ compilerOptions: { declaration: true } })
  files.forEach(file => {
    return prj.addSourceFileAtPath(file)
  })
  // 单测阶段
  const indexTs = path.resolve(__dirname, '../server/index.ts')
  // 编译后
  const indexDTs = path.resolve(__dirname, '../server/index.d.ts')
  const serverSf = prj.addSourceFileAtPath(
    existsSync(indexDTs) ? indexDTs : indexTs
  )
  const rpcServiceDef = serverSf.getFunction('RPCService')
  const rpcMethodDef = serverSf.getFunction('RPCMethod')
  if (rpcServiceDef == null || rpcMethodDef == null) throw Error('Cannot find RPCService or RPCMethod definition')

  const refedClasses = rpcServiceDef?.findReferencesAsNodes()
    // symbol <- callexpress <- decorator <- class
    .map(ref => ref.getParent()?.getParent()?.getParent())
    .filter(n => Node.isClassDeclaration(n)) as ClassDeclaration[]

  const genSf = prj.createSourceFile(path.resolve(__dirname, '_protocol-file-memory_.ts'), {
  }, { overwrite: true })

  const expInter = genSf.insertInterface(0, { name: 'ServiceCollection' })
  expInter.setIsExported(true)

  const rpcMetaData: TRPCMetaData = []
  refedClasses.forEach((c) => {
    const className = c.getName()
    if (className == null) throw Error('RPCService must be applied to a named class')
    const methods = findRPCMethods(c, rpcMethodDef)
    if (methods.length === 0) {
      throw Error(`RPCService(${className}) must have at least one method`)
    }
    rpcMetaData.push({
      name: className,
      path: c.getSourceFile().getFilePath(),
      methods: methods.map(m => ({ name: m.getName() }))
    })
    // 将 class 转换为 interface，模拟 rpc 的 protocol 声明
    const inter = genSf.addInterface({ name: className })
    inter.addJsDocs(c.getJsDocs().map(doc => doc.getStructure()))
    inter.addMethods(
      methods.map(m => {
        const ms = m.getSignature().getDeclaration() as MethodSignature
        let rtText = ms.getReturnType().getText()
        // 远程调用，返回值都是 Promise
        if (!/^Promise<.+>$/.test(rtText)) {
          rtText = `Promise<${rtText}>`
        }
        ms.setReturnType(rtText)
        return ms.getStructure()
      })
    )

    methods.map(m => collectMethodTypeDeps(m))
      .flat()
      .forEach(it => {
        // 添加 method 依赖的类型，否则无法编译通过
        if (it instanceof InterfaceDeclaration) {
          genSf.insertInterface(1, it.getStructure())
        } else if (it instanceof TypeAliasDeclaration) {
          genSf.insertTypeAlias(1, it.getStructure())
        } else {
          // TODO: error
        }
      })

    expInter.addProperty({ name: className, type: className })
  })

  // 疑似 ts-morph 的 bug，会出多许多重复的 interface Promise
  genSf.getInterfaces().forEach(it => {
    if (it.getName() === 'Promise') {
      it.remove()
    }
  })

  const protocolFileContent = genSf.getEmitOutput({ emitOnlyDtsFiles: true })
    .getOutputFiles().map(file => file.getText())
    .join('\n')
  return {
    dts: protocolFileContent,
    meta: rpcMetaData
  }
}

function findRPCMethods (service: ClassDeclaration, rpcMethodDef: FunctionDeclaration): MethodDeclaration[] {
  return service.getMethods()
    .filter(m => m.getDecorators().some(
      d => d.getNameNode()
        .getDefinitionNodes()
        .some(n => n === rpcMethodDef)
    ))
}

type ITDeclaration = InterfaceDeclaration | TypeAliasDeclaration
export function collectMethodTypeDeps (method: MethodDeclaration): ITDeclaration[] {
  return [...method.getParameters(), method.getReturnTypeNodeOrThrow()]
    .map(n => collectTypeDeps(n))
    .flat()
}

export function collectTypeDeps (t: Node): ITDeclaration[] {
  const typeRefDeclarationSet = new Set<ITDeclaration>()

  if (t instanceof TypeReferenceNode) {
    findITDeclaration(t)
  } else if (isITDeclaration(t)) {
    typeRefDeclarationSet.add(t)
  }

  // 深度优先遍历树，找到引用Type类型，然后找到 Declaration
  queryInTree(t)

  return Array.from(typeRefDeclarationSet)

  function queryInTree (n: Node): void {
    n.forEachChild(c => {
      if (c instanceof TypeReferenceNode) findITDeclaration(c)

      queryInTree(c)
    })
  }

  function isITDeclaration (n: Node): n is ITDeclaration {
    return n instanceof InterfaceDeclaration ||
      n instanceof TypeAliasDeclaration
  }

  function findITDeclaration (n: Node): void {
    n.getChildrenOfKind(SyntaxKind.Identifier)
      .map(i => i.getSymbol())
      .flat()
      .map(s => s?.getDeclarations())
      .flat()
      .forEach(d => {
        if (d != null && isITDeclaration(d)) typeRefDeclarationSet.add(d)
      })
  }
}
