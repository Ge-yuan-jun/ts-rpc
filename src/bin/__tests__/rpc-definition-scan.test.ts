import path from 'path'
import { MethodDeclaration, Project } from 'ts-morph'
import { collectMethodTypeDeps, collectTypeDeps, scan } from '../rpc-definition-scan'

test('scan', () => {
  const { dts, meta } = scan([path.resolve(__dirname, '*.ts')])

  expect(dts).toMatchSnapshot()
  expect(meta).toEqual([{
    name: 'User',
    path: expect.stringContaining('/src/bin/__tests__/controller-example.ts'),
    methods: [
      { name: 'getInfoById1' },
      { name: 'getInfoById2' },
      { name: 'getInfoById3' },
      { name: 'getUnreadMsg' }
    ]
  }, {
    name: 'Foo',
    path: expect.stringContaining('/src/bin/__tests__/controller-example.ts'),
    methods: [{ name: 'bar' }]
  }])
})

test('collectMethodTypeDeps', () => {
  const prj = new Project({ compilerOptions: { declaration: true } })
  const sf = prj.createSourceFile('test.ts', `
    interface A {}
    type B {}
    interface Z {}

    class Test {
      foo (a: A, b: B, c: string): Z {}
    }
  `)

  const deps = collectMethodTypeDeps(
    sf.getClass('Test')?.getMethod('foo') as MethodDeclaration,
    prj
  )
  expect(deps.map(d => d.getName())).toEqual(['A', 'B', 'Z'])
})

test('collectTypeDeps', () => {
  const prj = new Project({ compilerOptions: { declaration: true } })
  const sf = prj.createSourceFile('test.ts', `
    interface A {
      b: B
    }
    interface B {
      a: A
    }

    interface X {
      a: A
      b: B
      ab: A | B
      s: string
      n: number
      f: () => any
    }
    type Y = A | B | null | undefined
  `)
  const iDeps = collectTypeDeps(sf.getInterfaceOrThrow('X'), prj)
  expect(iDeps.map(i => i.getName())).toEqual(['X', 'A', 'B'])

  const tDeps = collectTypeDeps(sf.getTypeAliasOrThrow('Y'), prj)
  expect(tDeps.map(t => t.getName())).toEqual(['Y', 'A', 'B'])
})
