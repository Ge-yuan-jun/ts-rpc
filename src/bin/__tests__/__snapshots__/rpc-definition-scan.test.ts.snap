// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scan 1`] = `
"export interface ServiceCollection {
    User: User;
    Foo: Foo;
}

type TUserInfo = CUserInfo & IUserInfo;

interface IUserInfo {
    name: string;
    age: Date;
    avatar: string;
}

class CUserInfo {
    
    name: string;
    age: Date;
    avatar: string;
}

/**
 * service doc
 */
interface User {
    /**
     * method doc
     */
    getInfoById1(id: string): Promise<CUserInfo>;
    getInfoById2(id: string): Promise<IUserInfo>;
    getInfoById3(id: string): Promise<TUserInfo>;
    getUnreadMsg(id: string): Promise<string[]>;
}

interface Foo {
    bar(): Promise<void>;
}
"
`;
